sycl/include/CL/sycl/detail/program_impl.hpp

```diff
@@ -89,7 +95,15 @@ class program_impl {
     CHECK_OCL_CODE(clGetProgramInfo(ClProgram, CL_PROGRAM_DEVICES,
                                     sizeof(cl_device_id) * NumDevices,
                                     ClDevices.data(), nullptr));
-    Devices = vector_class<device>(ClDevices.begin(), ClDevices.end());
+    vector_class<device> SyclContextDevices = Context.get_devices();
+    std::map<cl_device_id, device> IdToDeviceMap;
+    for (auto &Dev : SyclContextDevices) {
+      IdToDeviceMap[Dev.get()] = Dev;
+    }
```

Comment:
It is not clear to me whether when you are calling `sort_devices_by_cl_device_id()` you have gone always through here. If so, since you are using `std::map<cl_device_id, device>`, if you are iterated on it it is already sorted by `cl_device_id`

Reply:
The map is used only in the interoperability constructor, whereas the sorting_by_cl_device_id() for device check is called in the program link constructor. The programs being linked may or may not constructed through the interoperability constructor. 

I don't know if std::sort is optimized for already sorted containers but I can add an if(!is_sorted(sortedDevices.begin(),sortedDevices.end()) check before calling the sort method. I wonder if it is it really important to do so.

Reply:
@keryell Thank you for the review. The review is also in progress here: https://github.com/otcshare/llvm/pull/53 
Apologies for the confusion. I didn't expect to receive comments from those other than I requested.

Reply:
@keyrell You may continue to comment here but I intend to close this pull request after the comments here are resolved. I will push changes to this branch after the pull request here https://github.com/otcshare/llvm/pull/53 receives approval.

Reply:
No I do not think this micro-optimization is important.
But thank you for the clarification.
And since you have it at hand, feel free to add it as a comment somewhere.
Comments are super useful for reviewers and generally for any non write-only code. ;-)
Thanks.